Change QPDF::copyForeignObject to return a null object when called wi…

m-holger opened the pull request:
…th a /Pages object (fixes #1011)

If copyForeignObject is called with a pages object, the object will not be inserted into the obj_copier.obj_map during normal processing. In the return statement the missing entry is default constructed, with the consequence that an uninitialized object handle is returned. All future references to the object are also replaced with an uninitialized object rather than a null object.

m-holger commented:
There probably should be a test case for this.

jberkenbilt commented:
Is this made obsolete by #1028?

m-holger commented:
No. This one deals with severely damaged files with garbage objects pretending to be page tree nodes

jberkenbilt commented:
I agree that there should be a test for this. Do we know how to reproduce it? Anyway, the new code is clearly better than the old code, so I'm inclined to merge it, though I'm worried that taking the change will cover up some other bug. I think I need to understand better when this happens. I think it would be better if we never inserted uninitialized objects rather than detecting them and returning null instead.

m-holger commented:
> Do we know how to reproduce it? 

All that is required is to change the /Type entry of a top level foreign object, e.g. the /AcroForm dictionary, to /Pages. 
That will result in the ObjGen not being inserted into obj_copier.object_map. On trying to retrieve the missing object from the object_map it gets default constructed as an uninitialized object handle. Similarly, deliberately trying to foreign copy a /Pages  object will have the same result.

m-holger commented:
> Is this made obsolete by https://github.com/qpdf/qpdf/pull/1028?

> No. This one deals with severely damaged files with garbage objects pretending to be page tree nodes

Sorry, it's been a while since I did this. This PR is actually the first commit in #1028

m-holger commented:
I've added a test.

>  I think it would be better if we never inserted uninitialized objects rather than detecting them and returning null instead.

I thought it was alright given that the check/fix was at the point of default creation but I've updated to move the check to before the map access.

jberkenbilt commented on libqpdf/QPDF.cc:
```diff
@@ -2034,13 +2034,13 @@ QPDF::copyForeignObject(QPDFObjectHandle foreign)
-        result = QPDFObjectHandle::newNull();
+    auto og = foreign.getObjGen();
+    if (!obj_copier.object_map.count(og)) {
         warn(damagedPDF("Unexpected reference to /Pages object while copying foreign object. "
```
I missed this before, but to be consistent with the style of most warnings, you could you change this to lower-case and semicolon-separated, as in `unexpected reference to /Pages object while copying foreign object;  replacing with null`

jberkenbilt reviewed (changes requested):
Great, thanks. Requesting a small change to the warning that I overlooked in my previous review.

jberkenbilt commented:
I'm stepping away for a few hours, but if this is not updated when I return, I will merge as is and tweak the warning message in a separate commit.
